AstarteDeviceSDKCSharp/Data/AstarteFailedMessageStorage.cs

```diff
@@ -18,170 +18,323 @@
+                {
+                    continue;
+                }
+
+
+                if (!IsMessageExist(message.Id))
+                {
+                    retention = message.ApplicationMessage.UserProperties.Where(x => x.Name == "Retention").FirstOrDefault();
+
+                    if (retention == null)
+                    {
+                        continue;
+                    }
+
+                    if (retention.Value == "STORED")
+                    {
+                        if (message.ApplicationMessage.MessageExpiryInterval > 0)
+                        {
+                            await InsertStored
+                              (
+                                  message.ApplicationMessage.Topic,
+                                  message.ApplicationMessage.Payload,
+                                  (int)message.ApplicationMessage.QualityOfServiceLevel,
+                                  message.Id,
+                                  (int)message.ApplicationMessage.MessageExpiryInterval
+                              );
+                            stored.Add(message);
+                        }
+                        else
+                        {
+                            await InsertStored
+                              (
+                                  message.ApplicationMessage.Topic,
+                                  message.ApplicationMessage.Payload,
+                                  (int)message.ApplicationMessage.QualityOfServiceLevel,
+                                  message.Id
+                              );
+                            stored.Add(message);
+                        }
+                    }
+                    else if (retention.Value == "VOLATILE")
+                    {
+
+                        if (message.ApplicationMessage.MessageExpiryInterval > 0)
+                        {
+                            InsertVolatile
+                            (
+                                message.ApplicationMessage.Topic,
+                                message.ApplicationMessage.Payload,
+                                (int)message.ApplicationMessage.QualityOfServiceLevel,
+                                message.Id,
+                                (int)message.ApplicationMessage.MessageExpiryInterval
+                            );
+                            stored.Add(message);
+                        }
+                        else
+                        {
+                            InsertVolatile
+                            (
+                                message.ApplicationMessage.Topic,
+                                message.ApplicationMessage.Payload,
+                                (int)message.ApplicationMessage.QualityOfServiceLevel,
+                                message.Id
+                            );
+                            stored.Add(message);
+                        }
+                    }
+                }
```

Comment:
I think you can improve this code by checking if the retention.Value == "DISCARD", and after that you can put I single `stored.Add(message)` at the end of if else if block instead of put four times store.add in each specific if block.
```suggestion
                    if (retention == null ||  retention.Value == "DISCARD")
                    {
                        continue;
                    }

                    if (retention.Value == "STORED")
                    {
                        if (message.ApplicationMessage.MessageExpiryInterval > 0)
                        {
                            await InsertStored
                              (
                                  message.ApplicationMessage.Topic,
                                  message.ApplicationMessage.Payload,
                                  (int)message.ApplicationMessage.QualityOfServiceLevel,
                                  message.Id,
                                  (int)message.ApplicationMessage.MessageExpiryInterval
                              );                            
                        }
                        else
                        {
                            await InsertStored
                              (
                                  message.ApplicationMessage.Topic,
                                  message.ApplicationMessage.Payload,
                                  (int)message.ApplicationMessage.QualityOfServiceLevel,
                                  message.Id
                              );
                        }
                    }
                    else if (retention.Value == "VOLATILE")
                    {

                        if (message.ApplicationMessage.MessageExpiryInterval > 0)
                        {
                            InsertVolatile
                            (
                                message.ApplicationMessage.Topic,
                                message.ApplicationMessage.Payload,
                                (int)message.ApplicationMessage.QualityOfServiceLevel,
                                message.Id,
                                (int)message.ApplicationMessage.MessageExpiryInterval
                            );
                        }
                        else
                        {
                            InsertVolatile
                            (
                                message.ApplicationMessage.Topic,
                                message.ApplicationMessage.Payload,
                                (int)message.ApplicationMessage.QualityOfServiceLevel,
                                message.Id
                            );                         
                        }
                        stored.Add(message);
                    }
                }
```